falcon/typing.py

```diff
@@ -11,19 +11,50 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
 """Shorthand definitions for more complex types."""
 
-from typing import Any, Callable, Pattern, Union
+from __future__ import annotations
+
+import http
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    MutableMapping,
+    Optional,
+    Pattern,
+    Tuple,
+    TYPE_CHECKING,
+    Union,
+)
+
+if TYPE_CHECKING:
+    from typing import Protocol
+
+    from falcon.request import Request
+    from falcon.response import Response
+
+    class Serializer(Protocol):
+        def serialize(
+            self,
+            media: MutableMapping[str, Union[str, int, None, Link]],
+            content_type: str,
+        ) -> bytes: ...
+
+    class MediaHandlers(Protocol):
+        def _resolve(
+            self, media_type: str, default: str, raise_not_found: bool = False
+        ) -> Tuple[Serializer, Optional[Callable], Optional[Callable]]: ...
+
 
-from falcon.request import Request
-from falcon.response import Response
+Link = Dict[str, str]
 
 # Error handlers
-ErrorHandler = Callable[[Request, Response, BaseException, dict], Any]
+ErrorHandler = Callable[['Request', 'Response', BaseException, dict], Any]
```

Comment:
Why cannot we import `Request` and `Response` here, is it because of cyclic references? We could import just under `TYPE_CHECKING` if that helps.

Reply:
Importing under type_checking requires strings in an union like this, since it's in the right side and it's evaluated